README.md
----------------------------------------------------------------------------------------
OvGME
Enhanced Generic Mod Enabler and manager.
----------------------------------------------------------------------------------------
**Version: 1.8.0, May 2022**
- increased repository download speed (limited only(?) by the repo server and your ISP)
- no installer anymore, sorry.
- based on the last version released some years ago

----------------------------------------------------------------------------------------

Home page: https://github.com/VNAO-Solid/ovgme

Credits:
----------------------------------------------------------------------------------------
All credits to Sedenion (Eri M.?). He created this wonderful tool afaik.

He is working on a new mod manager called "Open Mod Manager" (https://github.com/sedenion/OpenModMan).
Check it out regularly in order not to miss switching from OvGME to this new tool!

Presentation
----------------------------------------------------------------------------------------
OvGME is a Mod manager based on the idea and concept of JSGME, it take the GME acronym
from JSGME which stands for Generic Mod Enabler. The main purpose of OvGME is to provide an
easy way to import and enable mods for games then restore original files when disabling mods.
OvGME works by comparing the destination folder file tree with the given mod file tree,
then identifies what files to replace, create and save as backup.

OvGME implements the following key features:

- Support for multiple destination folder (game, software, etc.) through the main GUI.
- Custom Mods and Backup folder for each destination folder.
- Destination folder file tree snapshot and comparison based on xxHash.
- Zipped Mods files (Mod-Archives).
- Mods description (readme.txt).
- Mods versioning.
- Mods enabling profile for each destination folder.
- Mods network repositories.
- Mods online updating and downloading.
- Zip Mod-Archive creation tool.
- XML repository creation tool.
- Detailed embedded Help.


If you already use JSGME
Please read carefully the "OvGME and JSGME" chapter in the embedded help before doing
anything.

Screenshots
----------------------------------------------------------------------------------------
![OvGME Screenshot](snaps/screenshot.png)

Third party libraries
----------------------------------------------------------------------------------------


OvGME >=v1.8.0 uses curl to download files:

CURL COPYRIGHT AND PERMISSION NOTICE

Copyright (c) 1996 - 2022, Daniel Stenberg, daniel@haxx.se, and many contributors, see
the THANKS file.

All rights reserved.

Permission to use, copy, modify, and distribute this software for any purpose with or
without fee is hereby granted, provided that the above copyright notice and this permission
notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.

Except as contained in this notice, the name of a copyright holder shall not be used in
advertising or otherwise to promote the sale, use or other dealings in this Software
without prior written authorization of the copyright holder.



CHANGELOG.txt
OvGME 1.8.0, 2022-05-05

- Use libcurl to speed up downloads


OvGME 1.7.4, 2017-13-11

- Fix crash when pressing Esc key in configuration window.
- Config combo list is now alphabetically sorted.
- Fix window restored as minimized at startup.

OvGME 1.7.3, 2017-01-08

- Fix Zip files created using Windows Explorer not recognized as valid Mod-Archive.
- Snapshot comparison log output speed optimization.
- Fix make mod-archive overwrite cancel process.

OvGME 1.7.2, 2017-01-06

- Fix mod description display when using arrow keys.
- Add "Sort by enabled" option for mods list view.
- Add reposition of mods list scroll for long list.

OvGME 1.7.1, 2017-20-02

- Added support for multiple mods profile per config.
- Fixed version column width for large version numbers.
- Improving logs reports.
- New debug logs access via GUI.
- Help update

OvGME 1.7.0, 2017-14-02

- Dialog windows are now resizable.
- Main window size and position saved and restored.
- Global GUI improvement, more clear, better readability.
- Help updated and corrected.

OvGME 1.6.9, 2017-11-02

- Add Edit button on main dialog.
- Add Root path entry on main dialog.
- Fix some GUI elements.

OvGME 1.6.8, 2017-25-01

- Fixed config edit backup dir not updating which produce backup corruption.
- Added warning confirmation message if backup folder is already used.

OvGME 1.6.7, 2017-18-01

- Fixed download size cropped to 2147483647 bytes.
- Fixed XML repository parsing error.
- Added XML parsing test feature.

OvGME 1.6.6, 2016-21-11

- Fixed crash on read error or bad alloc during snapshot.

OvGME 1.6.5, 2016-31-08

- Fixed XML source generator segfault.
- Added ability to keeps configuration files during uninstall.
- Fixed uninstall process don't clean backup data.
- Fixed some dialog text spelling and sentences.
- "Game" is now a generic "Configuration".
- Help updated and corrected.

OvGME 1.6.4, 2016-28-08

- Added delete Mod on right-click menu
- Added quick make Mod-Archive on right-click menu
- Fixed need "Run As administrator" to write files

OvGME 1.6.3, 2016-26-08

- Fixed Make Mod-Archive bug when created in Mod stock folder
- Fixed bad optimization on Mod overlap check
- Fixed error on read or write 0 bytes
- Fixed main progress bar not working
- Fixed Directory-Mod has sam version than Mod-Archive
- Fixed some Cancel button logic
- Added compression level option for Make Mod-Archive
- Fixed Undo Mod inconsistent warning log output
- Backup cleaning optimization
- Added Mod Process safe Cancel and Undo logic

OvGME 1.6.2, 2016-20-08

- Mod download Url string corruption fixed
- Mod Apply/Backup logic modification
- Some correcting on help
- Added some log output

OvGME 1.6.1, 2016-20-08

- Per-User config and backups fixed.
- Ability to change game or config during Mods processing fixed.
- Added Mods processing cancel button for safe cancel.
- Added confirmation message for Mods profile overwrite.
- Added log output for network, write and read errors.

OvGME 1.6, 2016-20-08

- Fixed bug for already encoded URL being re-encoded with "%" char interpreted.
- Added support for HTTP chunked transfer.
- Added some error check for XML parsing.
- Empty error message on repository check fail fixed.
- Help updated and corrected.
- Help now starts with good window size.
- Added help source files.

OvGME 1.5.9, 2016-19-08

- Rewritten network code to avoid threading sync bugs.
- Better network error handling.
- Reworked Repositories feature.
- Added Ability to enable or disable repository.
- Help and Readme updated.

OvGME 1.5.1, 2016-15-08

- Jumping to version 1.5.
- New code architecture with headers and sources.
- Added Mod versioning support.
- Added network repository feature in alpha stage.

OvGME 1.2.1, 2016-11-08

- Fixed two ugly memory leaks.
- New write error check for mod apply and and backup.
- Write performances slightly improved.
- Auto text wrap disabled for snapshot output log.
- New snapshot cheksum method based on xxHash (faster).
- Source code: most functions renamed for better naming logic.
- Help corrected.

OvGME 1.2, 2016-10-08

- Added game snapshot feature (CRC-32 based).
- Added Mods profile feature.
- New threaded method for mod apply/restore to avoid "Not Responding" during heavy operations.
- Overall mod apply/restore algorithm optimized and reinforced (faster, less memory usage).
- Main menu logic improved (unavailable commands are grayed).
- Overall memory management optimization.
- Progress bar logic improved (now work correctly).
- Cleaned code & several optimizations.
- Help updated

OvGME 1.1, 2016-08-08

- Adding support for generic 'description.txt' or 'readme.txt' as description text file for mod archives.
- Adding right-click pop-up menu for mod list with "open folder or archive" option for quick access to mod content.
- Adding ability to choose a custom backup directory for a game.
- Adding ability to choose absolute path for mods stock directory for a game.
- Adding buttons to enable and disable mods in main dialog frame.
- New mod list view with visible enable/disable icon.

OvGME 1.0, 2016-07-08

- First release


src/gme_dlg_main.cpp
/*
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/
 */

#include "gme.h"
#include "gme_tools.h"
#include "gme_game.h"
#include "gme_conf.h"
#include "gme_mods.h"
#include "gme_prof.h"
#include "gme_repo.h"
#include "gme_netw.h"
#include "gme_logs.h"

#include "gme_dlg_about.h"
#include "gme_dlg_debug.h"
#include "gme_dlg_repo.h"
#include "gme_dlg_snap.h"
#include "gme_dlg_mods.h"
#include "gme_dlg_prof.h"
#include "gme_dlg_game.h"
#include "gme_dlg_main.h"
#include <ctime>

//#define TEST_VERSION

/*
  function for main init, startup process
*/
void GME_MainInit()
{
  GME_Logs(GME_LOG_NOTICE, "GME_MainInit", "OvGME Initialization", "Log history start");

  GME_ConfLoadCfg();

  /* create main icon */
  g_hicnMain = (HICON)LoadImage(g_hInst, MAKEINTRESOURCE(IDI_ICON),IMAGE_ICON,0,0,LR_DEFAULTSIZE|LR_SHARED);

  /* create main menu */
  g_hmnuMain = LoadMenu(NULL, MAKEINTRESOURCE(IDR_MENU));

  /* get profile submenu */
  g_hmnuSubProf = GetSubMenu(GetSubMenu(g_hmnuMain, 1), 3);

  /* list view image list */
  HIMAGELIST hImgLst;
  HBITMAP hBitmap;
  hImgLst = ImageList_Create( 16, 16, ILC_COLOR32, 3, 0 );
  hBitmap = (HBITMAP)LoadImage(g_hInst, MAKEINTRESOURCE(IDB_MODDIR), IMAGE_BITMAP, 16, 16, 0);
  ImageList_Add(hImgLst, hBitmap, NULL); DeleteObject(hBitmap);
  hBitmap = (HBITMAP)LoadImage(g_hInst, MAKEINTRESOURCE(IDB_MODZIP), IMAGE_BITMAP, 16, 16, 0);
  ImageList_Add(hImgLst, hBitmap, NULL); DeleteObject(hBitmap);
  hBitmap = (HBITMAP)LoadImage(g_hInst, MAKEINTRESOURCE(IDB_MODENABLED), IMAGE_BITMAP, 16, 16, 0);
  ImageList_Add(hImgLst, hBitmap, NULL); DeleteObject(hBitmap);
  SendMessageW(GetDlgItem(g_hwndMain, LVM_MODSLIST), LVM_SETIMAGELIST, LVSIL_SMALL, (LPARAM)hImgLst);
  SendMessageW(GetDlgItem(g_hwndMain, LVM_MODSLIST), LVM_SETIMAGELIST, LVSIL_NORMAL, (LPARAM)hImgLst);
  DeleteObject(hImgLst);

  /* set list window-style */
  if(!GME_ConfGetEnSort()) {
    int dwstyle = GetWindowLongPtrA(GetDlgItem(g_hwndMain, LVM_MODSLIST), GWL_STYLE);
    SetWindowLong(GetDlgItem(g_hwndMain, LVM_MODSLIST), GWL_STYLE, dwstyle|LVS_SORTASCENDING);
  } else {
    SendMessageW(GetDlgItem(g_hwndMain, CHK_SORTMODS), BM_SETCHECK, 1, 1);
  }

  /* set style for modlist */
  SendMessageW(GetDlgItem(g_hwndMain, LVM_MODSLIST), LVM_SETEXTENDEDLISTVIEWSTYLE, 0, LVS_EX_FULLROWSELECT|LVS_EX_SUBITEMIMAGES);

  /* create colums in list view */
  wchar_t buff[64];
  LVCOLUMNW lvcol;
  std::string info_msg;
  memset(&lvcol, 0, sizeof(LV_COLUMNW));
  lvcol.mask = LVCF_TEXT|LVCF_WIDTH;
  lvcol.cchTextMax = 255;

  wcscpy(buff, L"Mod");
  lvcol.pszText = buff;
  lvcol.cx = 650;
  lvcol.iSubItem = 0;
  SendMessageW(GetDlgItem(g_hwndMain, LVM_MODSLIST), LVM_INSERTCOLUMNW, 0, (LPARAM)&lvcol);

  wcscpy(buff, L"Version");
  lvcol.pszText = buff;
  lvcol.cx = 70;
  lvcol.iSubItem = 1;
  SendMessageW(GetDlgItem(g_hwndMain, LVM_MODSLIST), LVM_INSERTCOLUMNW, 1, (LPARAM)&lvcol);

  HFONT Lucida = CreateFont(12,0,0,0,FW_DONTCARE,FALSE,FALSE,FALSE,DEFAULT_CHARSET,OUT_OUTLINE_PRECIS,
                          CLIP_DEFAULT_PRECIS,CLEARTYPE_QUALITY,VARIABLE_PITCH,TEXT("Lucida Console"));

  SendMessage(GetDlgItem(g_hwndMain, ENT_MODDESC), WM_SETFONT, (WPARAM)Lucida, 1);

  SetWindowPos(g_hwndMain, NULL, GME_ConfGetWinX(), GME_ConfGetWinY(), GME_ConfGetWinW(), GME_ConfGetWinH(), SWP_NOZORDER);

  #ifdef TEST_VERSION
  time_t ttime = time(0);
  tm *local_time = localtime(&ttime);

  info_msg = "TEST VERSION -- TEST VERSION -- TEST VERSION\r\n\r\n";
  info_msg += "Report any issues to VFA-146 Solid on VNAO or CVW-14 discord\r\n\r\n";

  if (1900+local_time->tm_year > 2021 && local_time->tm_mon+1 > 5)
  {
    info_msg += "This version expired on 2022-05-31 ";
    GME_DialogInfo(g_hwndMain, GME_StrToWcs(info_msg));
    exit(1);
  }
  else
  {
    info_msg += "This version expires on 2022-05-31 ";
    GME_DialogInfo(g_hwndMain, GME_StrToWcs(info_msg));
  }

  #endif // TEST_VERSION
}

/*
  function to create pop menu when right-click on LV mods list
*/
void GME_MainPopMenuMods()
{
  /* mouse cursor position */
  POINT mouse_p;
  GetCursorPos(&mouse_p);


  HWND hlv = GetDlgItem(g_hwndMain, LVM_MODSLIST);

  bool single = (1 == SendMessage(hlv, LVM_GETSELECTEDCOUNT, 0, 0));

  /* get single selected Mod type */
  int type = -1;

  if(single) {
    wchar_t name_buff[255];

    LV_ITEMW lvitm;
    memset(&lvitm, 0, sizeof(LV_ITEMW));
    lvitm.mask = LVIF_TEXT|LVIF_IMAGE;
    lvitm.cchTextMax = 255;
    lvitm.pszText = name_buff;

    unsigned c = SendMessageW(hlv, LVM_GETITEMCOUNT, 0, 0);
    for(unsigned i = 0; i < c; i++) {
      if(SendMessageW(hlv, LVM_GETITEMSTATE, i, LVIS_SELECTED)) {
        lvitm.iItem = i;
        SendMessageW(hlv ,LVM_GETITEMW, 0, (LPARAM)&lvitm);
        type = lvitm.iImage;
        break; // single selection
      }
    }
  }

  HMENU hMenuPopup = CreatePopupMenu();
  if(single) {
    AppendMenu(hMenuPopup, MF_STRING, POP_TOGGLE, "Toggle");
    AppendMenu(hMenuPopup, MFT_SEPARATOR, MFT_SEPARATOR, "---");
    AppendMenu(hMenuPopup, MF_STRING, POP_OPENDIR, "Explore...");
    AppendMenu(hMenuPopup, MF_STRING, POP_MODDELETE, "Delete...");
    AppendMenu(hMenuPopup, MFT_SEPARATOR, MFT_SEPARATOR, "---");
    AppendMenu(hMenuPopup, MF_STRING, POP_MODCREATE, "Make Mod-Archive...");
  } else {
    AppendMenu(hMenuPopup, MF_STRING, POP_TOGGLE, "Toggle selected");
    AppendMenu(hMenuPopup, MF_STRING, POP_ENABLE, "Enable selected");
    AppendMenu(hMenuPopup, MF_STRING, POP_DISABLE, "Disable selected");
  }

  switch(type)
  {
  case 0:
    break;
  case 1:
    EnableMenuItem(hMenuPopup, POP_MODCREATE, MF_GRAYED);
    break;
  case 2:
    EnableMenuItem(hMenuPopup, POP_MODCREATE, MF_GRAYED);
    EnableMenuItem(hMenuPopup, POP_MODDELETE, MF_GRAYED);
    EnableMenuItem(hMenuPopup, POP_OPENDIR, MF_GRAYED);
    EnableMenuItem(hMenuPopup, POP_MODEDIT, MF_GRAYED);
    break;
  }

  TrackPopupMenu(hMenuPopup, TPM_TOPALIGN|TPM_LEFTALIGN, mouse_p.x, mouse_p.y, 0, g_hwndMain, NULL);
}

/*
  function for main exit, cleanup and exit process
*/
void GME_MainExit()
{
  GME_ConfSaveCfg();
  GME_ModsClean();
  GME_GameClean();
  GME_RepoClean();
}

/*
  controls resize & placement routine for main dialog window
*/
void GME_DlgMainResize()
{
  HWND hwd = g_hwndMain;
  RECT cli;
  GetClientRect(hwd, &cli);

  SetWindowPos(GetDlgItem(hwd, BOX_MAIN), NULL, 10, 77, cli.right-20, cli.bottom-115, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_QUIT), NULL, cli.right-90, cli.bottom-30, 80, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, ENT_MODDESC), NULL, 20, cli.bottom-160, cli.right-40, 110, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, LAB_MODDESC), NULL, 20, cli.bottom-175, 150, 15, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_MODDIS), NULL, 20, cli.bottom-203, 100, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_MODENA), NULL, 120, cli.bottom-203, 100, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_MODCANCEL), NULL, cli.right-100, cli.bottom-203, 80, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, PBM_MODPROC), NULL, 225, cli.bottom-203, cli.right-330, 22, SWP_NOZORDER);

  SetWindowPos(GetDlgItem(hwd, LVM_MODSLIST), NULL, 20, 135, cli.right-40, cli.bottom-345, SWP_NOZORDER);

  /* modify colums in list view */
  LVCOLUMNW lvcol;
  memset(&lvcol, 0, sizeof(LV_COLUMNW));
  lvcol.mask = LVCF_WIDTH;
  lvcol.cx = cli.right-125;
  lvcol.iSubItem = 0;
  SendMessageW(GetDlgItem(hwd, LVM_MODSLIST), LVM_SETCOLUMN, 0, (LPARAM)&lvcol);
  lvcol.cx = 60;
  lvcol.iSubItem = 1;
  SendMessageW(GetDlgItem(hwd, LVM_MODSLIST), LVM_SETCOLUMN, 1, (LPARAM)&lvcol);

  SetWindowPos(GetDlgItem(hwd, CHK_SORTMODS), NULL, 20, 115, cli.right-40, 20, SWP_NOZORDER);

  SetWindowPos(GetDlgItem(hwd, BTN_IMPORTMOD), NULL, cli.right-105, 90, 85, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, ENT_MODSPATH), NULL, 20, 91, cli.right-130, 21, SWP_NOZORDER);

  SetWindowPos(GetDlgItem(hwd, BOX_TOP), NULL, 155, 0, cli.right-165, 80, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, CMB_GAMELIST), NULL, 165, 20, cli.right-260, 22, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_EDIGAME), NULL, cli.right-92, 19, 35, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, BTN_ADDGAME), NULL, cli.right-55, 19, 35, 23, SWP_NOZORDER);
  SetWindowPos(GetDlgItem(hwd, ENT_CFGRPATH), NULL, 165, 45, cli.right-185, 21, SWP_NOZORDER);
}


/*
  Mods list view automatic sorting enabled or not
*/
void GME_DlgMainSortList(bool byenabled)
{
  int dwstyle = GetWindowLongPtrA(GetDlgItem(g_hwndMain, LVM_MODSLIST), GWL_STYLE);

  if(byenabled) {
    dwstyle = dwstyle & ~LVS_SORTASCENDING;
    SetWindowLong(GetDlgItem(g_hwndMain, LVM_MODSLIST), GWL_STYLE, dwstyle);
  } else {
    SetWindowLong(GetDlgItem(g_hwndMain, LVM_MODSLIST), GWL_STYLE, dwstyle|LVS_SORTASCENDING);
  }
}

/*
  message callback for main dialog window
*/
BOOL CALLBACK GME_DlgMain(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
  switch(uMsg)
  {
  case WM_INITDIALOG:
    g_hwndMain = hwndDlg;

    /* global init */
    GME_MainInit();
    /* set main icon */
    SendMessage(g_hwndMain, WM_SETICON, ICON_BIG, (LPARAM)g_hicnMain);
    /* set main menu */
    SetMenu(hwndDlg, g_hmnuMain);
    /* disable cancel button */
    EnableWindow(GetDlgItem(g_hwndMain, BTN_MODCANCEL), false);
    /* Get game cfg and create list */
    if(!GME_GameUpdList()) {
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_WIZARD), hwndDlg, (DLGPROC)GME_DlgGameAdd);
    }
    GME_DlgMainResize();
    return true;

  case WM_SIZE:
    GME_DlgMainResize();
    return true;

  case WM_CLOSE:
    GME_MainExit();
    EndDialog(hwndDlg, 0);
    return true;

  case WM_NOTIFY:
    if(LOWORD(wParam) == LVM_MODSLIST) {
      if(((LPNMHDR)lParam)->code == NM_DBLCLK) {
        GME_ModsToggleSel(MODS_TOGGLE);
        return true;
      }
      if(((LPNMHDR)lParam)->code == NM_CLICK) {
        GME_ModsChkDesc();
        return true;
      }
      if(((LPNMHDR)lParam)->code == NM_RCLICK) {
        GME_MainPopMenuMods();
        return true;
      }
      if(((LPNMHDR)lParam)->code == LVN_ITEMCHANGED) {
        GME_ModsChkDesc();
        return true;
      }
    }
    return true;

  case WM_COMMAND:
    if(LOWORD(wParam) > 40100) {
      GME_ProfApply(LOWORD(wParam));
      return true;
    }
    switch(LOWORD(wParam))
    {
    case BTN_QUIT:
      GME_MainExit();
      EndDialog(hwndDlg, 0);
      return true;

    case CHK_SORTMODS:
      GME_DlgMainSortList(SendMessage(GetDlgItem(hwndDlg, CHK_SORTMODS), BM_GETCHECK, 0, 0));
      GME_ModsUpdList();
      return true;

    case MNU_GAMEADD:
    case BTN_ADDGAME:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_GAME_ADD), hwndDlg, (DLGPROC)GME_DlgGameAdd);
      return true;

    case MNU_MODIMPORT:
    case BTN_IMPORTMOD:
      GME_ModsImport();
      return true;

    case MNU_MODCREATE:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_MOD_MAKE), hwndDlg, (DLGPROC)GME_DlgModsMake);
      return true;

    case MNU_GAMEREM:
      GME_GameRemCurCfg();
      return true;

    case MNU_GAMEEDIT:
    case BTN_EDIGAME:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_GAME_EDIT), hwndDlg, (DLGPROC)GME_DlgGameEdit);
      return true;

    case MNU_MODENA:
    case BTN_MODENA:
    case POP_ENABLE:
      GME_ModsToggleSel(MODS_ENABLE);
      return true;

    case MNU_MODDIS:
    case BTN_MODDIS:
    case POP_DISABLE:
      GME_ModsToggleSel(MODS_DISABLE);
      return true;

    case MNU_MODENAALL:
      GME_ModsToggleAll(MODS_ENABLE);
      return true;

    case MNU_MODDISALL:
      GME_ModsToggleAll(MODS_DISABLE);
      return true;

    case BTN_MODCANCEL:
      GME_ModsProcCancel();
      return true;

    case MNU_SNAPCREATE:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_SNAP_NEW), hwndDlg, (DLGPROC)GME_DlgSnapNew);
      return true;

    case MNU_SNAPCOMPARE:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_SNAP_CMP), hwndDlg, (DLGPROC)GME_DlgSnapComp);
      return true;

    case MNU_PROFILSAVE:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_PROF_NEW), hwndDlg, (DLGPROC)GME_DlgProfNew);
      return true;

    case MNU_PROFILDELT:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_PROF_DEL), hwndDlg, (DLGPROC)GME_DlgProfDel);
      return true;

    case MNU_REPOSCONFIG:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_REPOS_LST), hwndDlg, (DLGPROC)GME_DlgRepConf);
      return true;

    case MNU_REPOSQUERY:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_REPOS_UPD), hwndDlg, (DLGPROC)GME_DlgRepUpd);
      return true;

    case MNU_REPOSMKXML:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_REPOS_XML), hwndDlg, (DLGPROC)GME_DlgRepXml);
      return true;

    case MNU_REPOSTSXML:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_REPOS_XTS), hwndDlg, (DLGPROC)GME_DlgRepXts);
      return true;

    case MNU_ABOUT:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_ABOUT), hwndDlg, (DLGPROC)GME_DlgAbout);
      return true;

    case MNU_HELP:
      ShellExecute(hwndDlg, _T( "open" ), _T( "ovgme.chm" ), NULL, NULL, SW_NORMAL );
      return true;

    case MNU_LOGS:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_LOGS), hwndDlg, (DLGPROC)GME_DlgDebug);
      return true;

    case CMB_GAMELIST:
      if(HIWORD(wParam) == CBN_SELCHANGE) {
        GME_GameChkSelect();
      }
      return true;

    case POP_TOGGLE:
      GME_ModsToggleSel(MODS_TOGGLE);
      return true;

    case POP_OPENDIR:
      GME_ModsExploreCur();
      return true;

    case POP_MODCREATE:
      DialogBox(g_hInst, MAKEINTRESOURCE(DLG_MOD_QMAKE), hwndDlg, (DLGPROC)GME_DlgModsQuickMake);
      return true;

    case POP_MODDELETE:
      GME_ModsDeleteCur();
      return true;
    }
  }

  return false;
}

